### PR TITLE
Refactor authentication logic

### LINKED ISSUE #502: Refactor authentication logic
There's one thing that bothers me with authentication logic. So here's a good task for you.

## Issue
Currently **all** of authentication logic implementation (as well as JWT creation logic) is stored inside a  `AccountsController.SignIn()` controller method:
https://github.com/ita-social-projects/WhatBackend/blob/899f397be4e51f81fc156f059c3ced0a0a60d669/CharlieBackend.Api/Controllers/AccountsController.cs#L123
which is a code smell.

## Solution
What we need to do:
- split JWT generation & the rest of authentication logic
- move the logic into corresponding services/classes (into abstraction level corresponding to them)
---
Also, what would be a good thing is replacing these hardcoded strings of claim types (Email, Id, etc) with corresponding constants values:
https://github.com/ita-social-projects/WhatBackend/blob/899f397be4e51f81fc156f059c3ced0a0a60d669/CharlieBackend.Api/Controllers/AccountsController.cs#L131
https://github.com/ita-social-projects/WhatBackend/blob/899f397be4e51f81fc156f059c3ced0a0a60d669/CharlieBackend.Api/Controllers/AccountsController.cs#L132

If you're gonna do that, then don't forget to update these claim types in `CurrentUserService`:
https://github.com/ita-social-projects/WhatBackend/blob/899f397be4e51f81fc156f059c3ced0a0a60d669/CharlieBackend.Business/Services/CurrentUserService.cs#L59
https://github.com/ita-social-projects/WhatBackend/blob/899f397be4e51f81fc156f059c3ced0a0a60d669/CharlieBackend.Business/Services/CurrentUserService.cs#L32

### PR DESCRIPTION
**Pull request by DemetriosKoras**

### CODE REVIEW

File: CharlieBackend.Api/Helpers/JWTGenerator.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+namespace CharlieBackend.Api.Helpers
+{
+    public class JWTGenerator : IJWTGenerator
+    {
+        private AuthOptions _authOptions;
+        public JWTGenerator(IOptions<AuthOptions> authOptions)
+        {
+            _authOptions = authOptions.Value;
+        }
```
Comment (andrii-developer) [inline CharlieBackend.Api/Helpers/JWTGenerator.cs]: Please add one empty line between methods.

File: CharlieBackend.Api/Startup.cs [hunk 2]
```diff
@@ -147,6 +148,8 @@ public void ConfigureServices(IServiceCollection services)
             services.Configure<SwaggerOptions>(c => c.SerializeAsV2 = true);
 
             services.AddSwaggerExamplesFromAssemblyOf<Startup>();
+
+            services.AddTransient<IJWTGenerator, JWTGenerator>();
```
Comment (andrii-developer) [inline CharlieBackend.Api/Startup.cs]: Please move this line to the `CompositionRoot.cs`

File: CharlieBackend.Business/Services/CurrentUserService.cs [hunk 3]
```diff
@@ -1,10 +1,12 @@
 ﻿using CharlieBackend.Business.Services.Interfaces;
+using CharlieBackend.Business.Helpers;
 using CharlieBackend.Core.Entities;
 using Microsoft.AspNetCore.Http;
 using System;
 using System.Linq;
 using System.Security.Claims;
 
+
```
Comment (andrii-developer) [inline CharlieBackend.Business/Services/CurrentUserService.cs]: Please remove the empty line.

File: CharlieBackend.Api/Helpers/JWTGenerator.cs (hunk 1 above)
Comment (DemetriosKoras) [reply CharlieBackend.Api/Helpers/JWTGenerator.cs]: 43c7965

File: CharlieBackend.Business/Services/CurrentUserService.cs (hunk 3 above)
Comment (DemetriosKoras) [reply CharlieBackend.Business/Services/CurrentUserService.cs]: 29404d5

File: CharlieBackend.Api/Startup.cs (hunk 2 above)
Comment (DemetriosKoras) [reply CharlieBackend.Api/Startup.cs]: bc2aae1

File: CharlieBackend.Business/Helpers/ClaimConstants.cs [hunk 4]
```diff
@@ -0,0 +1,10 @@
+﻿
+namespace CharlieBackend.Business.Helpers
+{
+    public class ClaimConstants
```
Comment (andrii-developer) [inline CharlieBackend.Business/Helpers/ClaimConstants.cs]: Please make the class static.

File: CharlieBackend.Business/Helpers/IJWTGenerator.cs [hunk 5]
```diff
@@ -0,0 +1,9 @@
+﻿using CharlieBackend.Core.DTO.Account;
+
+namespace CharlieBackend.Business.Helpers
+{
+    public interface IJWTGenerator
```
Comment (andrii-developer) [inline CharlieBackend.Business/Helpers/IJWTGenerator.cs]: According to conventions, it should be `IJwtGenerator`. Also, don't forget to update implementation and fields.

**Review by andrii-developer — CHANGES_REQUESTED** [review]
Waiting for addressing 2 comments.

File: CharlieBackend.Business/Helpers/ClaimConstants.cs (hunk 4 above)
Comment (DemetriosKoras) [reply CharlieBackend.Business/Helpers/ClaimConstants.cs]: 903ff8b

File: CharlieBackend.Business/Helpers/IJWTGenerator.cs (hunk 5 above)
Comment (DemetriosKoras) [reply CharlieBackend.Business/Helpers/IJWTGenerator.cs]: 0214670

File: CharlieBackend.Root/CompositionRoot.cs [hunk 6]
```diff
@@ -63,7 +64,7 @@ public static void InjectDependencies(IServiceCollection services, IConfiguratio
             services.AddScoped<IScheduledEventHandlerFactory, ScheduledEventHandlerFactory>();
             services.AddScoped<IHomeworkStudentService, HomeworkStudentService>();
             services.AddScoped<IEventsService, EventsService>();
-
+            services.AddScoped<IJwtGenerator, JwtGenerator>();
```
Comment (gingerbread-monster) [inline CharlieBackend.Root/CompositionRoot.cs:68]: Why add this service as a scoped and not as a singleton? (Same goes for other services)
There's no point in creating new service instances on each request for the most of these services.

Comment (DemetriosKoras) [reply CharlieBackend.Root/CompositionRoot.cs:68]: It seems not viable because of we might be getting new AuthOptions data each new scope.

**Review by andrii-developer — APPROVED** [review]

**Pull request merged**